generate/base.py

```diff
@@ -142,13 +142,20 @@ def main(
     fabric.print(f"Time to instantiate model: {time.perf_counter() - t0:.02f} seconds.", file=sys.stderr)
 
     t0 = time.perf_counter()
-    with lazy_load(checkpoint_path) as checkpoint:
-        model.load_state_dict(checkpoint.get("model", checkpoint), strict=quantize is None)
-    fabric.print(f"Time to load the model weights: {time.perf_counter() - t0:.02f} seconds.", file=sys.stderr)
+
+    if quantize:
```

Comment:
Why did you have to split the logic around `quantize`?

Reply:
Because just like in the finetune script, the quantization of loaded weights needs to happen before moving to the device.

Reply:
The error I get otherwise is:

```
python generate/base.py --precision bf16-true --quantize bnb.nf4
```

```
Traceback (most recent call last):
  File "/home/adrian/repositories/lit-gpt/generate/base.py", line 185, in <module>
    CLI(main)
  File "/home/adrian/.conda/envs/lit-gpt/lib/python3.10/site-packages/jsonargparse/_cli.py", line 85, in CLI
    return _run_component(component, cfg_init)
  File "/home/adrian/.conda/envs/lit-gpt/lib/python3.10/site-packages/jsonargparse/_cli.py", line 147, in _run_component
    return component(**cfg)
  File "/home/adrian/repositories/lit-gpt/generate/base.py", line 149, in main
    load_checkpoint(fabric, model, checkpoint_path, strict=(not quantize))
  File "/home/adrian/repositories/lit-gpt/lit_gpt/utils.py", line 521, in load_checkpoint
    model.load_state_dict(state_dict, strict=strict)
  File "/home/adrian/repositories/lightning/src/lightning/fabric/wrappers.py", line 144, in load_state_dict
    return self._original_module.load_state_dict(state_dict=state_dict, strict=strict)
  File "/home/adrian/.conda/envs/lit-gpt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 2153, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for GPT:
        size mismatch for lm_head.weight: copying a param with shape torch.Size([50688, 4096]) from checkpoint, the shape in current model is torch.Size([103809024, 1]).
        size mismatch for transformer.h.0.attn.attn.weight: copying a param with shape torch.Size([12288, 4096]) from checkpoint, the shape in current model is torch.Size([25165824, 1]).
```

Reply:
The error I get if I don't split the logic is:

```
python generate/base.py --precision bf16-true --quantize bnb.fp4
```

```
Traceback (most recent call last):
  File "/home/adrian/repositories/lit-gpt/generate/base.py", line 189, in <module>
    CLI(main)
  File "/home/adrian/.conda/envs/lit-gpt/lib/python3.10/site-packages/jsonargparse/_cli.py", line 85, in CLI
    return _run_component(component, cfg_init)
  File "/home/adrian/.conda/envs/lit-gpt/lib/python3.10/site-packages/jsonargparse/_cli.py", line 147, in _run_component
    return component(**cfg)
  File "/home/adrian/repositories/lit-gpt/generate/base.py", line 153, in main
    load_checkpoint(fabric, model, checkpoint_path, strict=(quantize is None))
  File "/home/adrian/repositories/lit-gpt/lit_gpt/utils.py", line 521, in load_checkpoint
    model.load_state_dict(state_dict, strict=strict)
  File "/home/adrian/repositories/lightning/src/lightning/fabric/wrappers.py", line 144, in load_state_dict
    return self._original_module.load_state_dict(state_dict=state_dict, strict=strict)
  File "/home/adrian/.conda/envs/lit-gpt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 2153, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for GPT:
        size mismatch for lm_head.weight: copying a param with shape torch.Size([50688, 4096]) from checkpoint, the shape in current model is torch.Size([103809024, 1]).
        size mismatch for transformer.h.0.attn.attn.weight: copying a param with shape torch.Size([12288, 4096]) from checkpoint, the shape in current model is torch.Size([25165824, 1]).
        size mismatch for transformer.h.0.attn.proj.weight: copying a param with shape torch.Size([4096, 4096]) from checkpoint, the shape in current model is torch.Size([8388608, 1]).
        size mismatch for transformer.h.0.mlp.fc.weight: copying a param with shape torch.Size([16384, 4096]) from checkpoint, the shape in current model is torch.Size([33554432, 1]).
        size mismatch for transformer.h.0.mlp.proj.weight: copying a param with shape torch.Size([4096, 16384]) from checkpoint, the shape in current model is torch.Size([33554432, 1]).
        size mismatch for transformer.h.1.attn.attn.weight: copying a param with shape torch.Size([12288, 4096]) from checkpoint, the shape in current model is torch.Size([25165824, 1]).
        size mismatch for transformer.h.1.attn.proj.weight: copying a param with shape torch.Size([4096, 4096]) from checkpoint, the shape in current model is torch.Size([8388608, 1]).
        size mismatch for transformer.h.1.mlp.fc.weight: copying a param with shape torch.Size([16384, 4096]) from checkpoint, the shape in current model is torch.Size([33554432, 1]).
        size mismatch for transformer.h.1.mlp.proj.weight: copying a param with shape torch.Size([4096, 16384]) from checkpoint, the shape in current model is torch.Size([33554432, 1]).
        size mismatch for transformer.h.2.attn.attn.weight: copying a param with shape torch.Size([12288, 4096]) from checkpoint, the shape in current model is torch.Size([25165824, 1]).
        size mismatch for transformer.h.2.attn.proj.weight: copying a param with shape torch.Size([4096, 4096]) from checkpoint, the shape in current model is torch.Size([8388608, 1]).
```